RecoLocalTracker/SiStripClusterizer/plugins/SiStripClusters2ApproxClusters.cc

```diff
@@ -27,25 +45,113 @@ class SiStripClusters2ApproxClusters : public edm::stream::EDProducer<> {
   edm::EDGetTokenT<edmNew::DetSetVector<SiStripCluster> > clusterToken;
 
   unsigned int maxNSat;
+
+  edm::InputTag beamSpot;
+  edm::EDGetTokenT<reco::BeamSpot> beamSpotToken;
+
+  edm::ESGetToken<TrackerGeometry, TrackerDigiGeometryRecord> tkGeomToken_;
+
+  edm::FileInPath fileInPath;
+  SiStripDetInfo detInfo;
+
+  edm::ESGetToken<ClusterShapeHitFilter, CkfComponentsRecord> csfToken_;
+
+  edm::ESGetToken<SiStripNoises, SiStripNoisesRcd> stripNoiseToken_;
+  edm::ESHandle<SiStripNoises> theNoise;
 };
 
 SiStripClusters2ApproxClusters::SiStripClusters2ApproxClusters(const edm::ParameterSet& conf) {
   inputClusters = conf.getParameter<edm::InputTag>("inputClusters");
   maxNSat = conf.getParameter<unsigned int>("maxSaturatedStrips");
 
   clusterToken = consumes<edmNew::DetSetVector<SiStripCluster> >(inputClusters);
+
+  beamSpot = conf.getParameter<edm::InputTag>("beamSpot");
+  beamSpotToken = consumes<reco::BeamSpot>(beamSpot);
+
+  tkGeomToken_ = esConsumes();
+
+  fileInPath = edm::FileInPath(SiStripDetInfoFileReader::kDefaultFile);
+  detInfo = SiStripDetInfoFileReader::read(fileInPath.fullPath());
+
+  csfToken_ = esConsumes(edm::ESInputTag("", "ClusterShapeHitFilter"));
+
+  stripNoiseToken_ = esConsumes();
+
   produces<edmNew::DetSetVector<SiStripApproximateCluster> >();
 }
 
-void SiStripClusters2ApproxClusters::produce(edm::Event& event, edm::EventSetup const&) {
+void SiStripClusters2ApproxClusters::produce(edm::Event& event, edm::EventSetup const& iSetup) {
   auto result = std::make_unique<edmNew::DetSetVector<SiStripApproximateCluster> >();
   const auto& clusterCollection = event.get(clusterToken);
 
+  edm::Handle<reco::BeamSpot> beamSpotHandle;
+  event.getByToken(beamSpotToken, beamSpotHandle);  // retrive BeamSpot data
+  reco::BeamSpot const* bs = nullptr;
+  if (beamSpotHandle.isValid())
+    bs = &(*beamSpotHandle);
+
+  const auto& tkGeom = &iSetup.getData(tkGeomToken_);
+  const auto& theFilter = &iSetup.getData(csfToken_);
+  const auto& theNoise = &iSetup.getData(stripNoiseToken_);
+
+  // parameters for peakFilter in StripSubClusterShapeTrajectoryFilter...
+  float MeVperADCStrip = 9.5665E-4;
+  double subclusterWindow_ = .7;
+  double seedCutMIPs_ = .35;
+  double seedCutSN_ = 7.;
+  double subclusterCutMIPs_ = .45;
+  double subclusterCutSN_ = 12.;
+
   for (const auto& detClusters : clusterCollection) {
     edmNew::DetSetVector<SiStripApproximateCluster>::FastFiller ff{*result, detClusters.id()};
 
-    for (const auto& cluster : detClusters)
-      ff.push_back(SiStripApproximateCluster(cluster, maxNSat));
+    unsigned int detId = detClusters.id();
+    const GeomDet* det = tkGeom->idToDet(detId);
+    double nApvs;
+    double stripLength;
+    nApvs = detInfo.getNumberOfApvsAndStripLength(detId).first;
+    stripLength = detInfo.getNumberOfApvsAndStripLength(detId).second;
+    double barycenter_ypos;
+    barycenter_ypos = 0.5 * stripLength;
```

Comment:
```suggestion
    double nApvs = detInfo.getNumberOfApvsAndStripLength(detId).first;
    double stripLength = detInfo.getNumberOfApvsAndStripLength(detId).second;
    double barycenter_ypos = 0.5 * stripLength;
```

to avoid useless copies.